test/run.sh

```diff
@@ -14,43 +14,103 @@
+        shift
+        ;;
+      -*)
+        echo "Uknown option '$1'"
+        exit 1
+        ;;
+      *)
+        echo "Uknown argument '$1'"
+        exit 1
+        ;;
+    esac
+  done
+}
+
 function cleanup() {
-  echo Stopping container..
-  docker stop $container_datalab $selenium_container
+  if [[ $CONTAINER_STARTED -ne 0 ]]; then
+    echo Stopping container..
+    docker stop $container_datalab $selenium_container
+  fi
   exit
 }
 
-# For travis, we do not care about interrupts and cleanup,
-# it will just waste time
-if [ -z "$TRAVIS" ]; then
-  trap cleanup INT EXIT SIGHUP SIGINT SIGTERM
-fi
-
-TESTS_HOME=$HOME/datalab_tests
-mkdir -p $TESTS_HOME
-
-echo Starting Datalab container..
-container_datalab=$(docker run -d \
-  --entrypoint="/datalab/run.sh" \
-  -p 127.0.0.1:8081:8080 \
-  -v $TESTS_HOME:/content \
-  -e "ENABLE_USAGE_REPORTING=false" \
-  datalab)
-
-echo Starting selenium container..
-selenium_container=$(docker run -d -p 4444:4444 --net="host" selenium/standalone-chrome)
-
-echo -n Polling on Datalab..
-until $(curl --output /dev/null --silent --head --fail http://localhost:8081); do
-  printf '.'
-  sleep 1
-done
-echo ' Done.'
-echo -n Polling on Selenium..
-until $(curl --output /dev/null --silent --head --fail http://localhost:4444/wd/hub); do
-  printf '.'
-  sleep 1
-done
-echo ' Done.'
-
-mocha ui/test.js notebook/test.js unittests/frontend.js
+function makeTestsHome() {
+  TESTS_HOME=$HOME/datalab_tests
+  mkdir -p $TESTS_HOME
+}
+
+function startContainers() {
+  CONTAINER_STARTED=1
+  echo Starting Datalab container..
+  container_datalab=$(docker run -d \
+    --entrypoint="/datalab/run.sh" \
+    -p 127.0.0.1:8081:8080 \
+    -v $TESTS_HOME:/content \
+    -e "ENABLE_USAGE_REPORTING=false" \
+    datalab)
+
+  echo Starting selenium container..
+  selenium_container=$(docker run -d -p 4444:4444 --net="host" selenium/standalone-chrome)
+
+  echo -n Polling on Datalab..
+  until $(curl --output /dev/null --silent --head --fail http://localhost:8081); do
+    printf '.'
+    sleep 1
+  done
+  echo ' Done.'
+  echo -n Polling on Selenium..
+  until $(curl --output /dev/null --silent --head --fail http://localhost:4444/wd/hub); do
+    printf '.'
+    sleep 1
+  done
+  echo ' Done.'
+}
+
+function runIntegrationTests() {
+  echo Running mocha notebook tests
+  $MOCHA $HERE/notebook/test.js
+  echo Running mocha ui tests
+  $MOCHA $HERE/ui/test.js
+}
+
+function runUnitTests() {
+  echo Running jasmine tests
+  $HERE/node_modules/jasmine/bin/jasmine.js
+}
+
+function main() {
+  parseOptions "$@"
+
+  # For travis, we do not care about interrupts and cleanup,
+  # it will just waste time
+  if [ -z "$TRAVIS" ]; then
+    trap cleanup INT EXIT SIGHUP SIGINT SIGTERM
+  fi
+
+  # Unit tests are fast, run them all first
+  runUnitTests
+
+  if [[ $RUN_INTEGRATION_TESTS -ne 0 ]]; then
+    makeTestsHome
+    startContainers
+  fi
+
+  if [[ $RUN_INTEGRATION_TESTS -ne 0 ]]; then
+    runIntegrationTests
```

Comment:
Done. There used to be something between these two, and I didn't notice they were the same when I moved that piece elsewhere.